### PR TITLE
Implement channel keys in RobustIRC.

### PR DESCRIPTION
**Pull request by danielkim802**
The following changes are introduced (copying behavior of UnrealIRCd):
- Add ability to set and remove channel key only by the  channel operator
- Non-operators are shown an error message saying they need to be operator to modify keys
- Attempting to set a key with no password is a no op

### CODE REVIEW

File: internal/ircserver/cmd_join.go [hunk 1]
```diff
@@ -56,6 +56,7 @@ func (i *IRCServer) cmdJoin(s *Session, reply *Replyctx, msg *irc.Message) {
 			c = &channel{
 				name:  channelname,
 				nicks: make(map[lcNick]*[maxChanMemberStatus]bool),
+				key:   key,
```
Comment (stapelberg) [inline internal/ircserver/cmd_join.go:59]: This change surprises me: when a channel does not exist yet, I think it should be created without a key, regardless of whether the user tried to use a key to join the channel.

Whether a channel has a key or not is determined by the 'k' channel mode, so setting the struct field “key” without adjusting the modes is inconsistent (but, again, I think we should probably just remove this line).

File: serialization_test.go [hunk 2]
```diff
@@ -59,6 +59,7 @@ func TestSerialization(t *testing.T) {
 	logs = appendLog(logs, `{"Id": {"Id": 7}, "Session": {"Id": 1}, "Type": 2, "Data": "JOIN #test"}`)
 	logs = appendLog(logs, `{"Id": {"Id": 8}, "Session": {"Id": 1}, "Type": 2, "Data": "MODE #test +i"}`)
 	logs = appendLog(logs, `{"Id": {"Id": 9}, "Session": {"Id": 1}, "Type": 2, "Data": "INVITE mero #test"}`)
+	logs = appendLog(logs, `{"Id": {"Id": 10}, "Session": {"Id": 1}, "Type": 2, "Data": "MODE #test +k 1234"}`)
```
Comment (stapelberg) [inline serialization_test.go:62]: (Why) is it necessary to update this test? We already have test coverage for the `MODE #channel +k` command in the other files, and this test seems unrelated?

**Review by stapelberg — CHANGES_REQUESTED** [review]
Currently, you have one commit per file that you update. Can you squash these commits together? We usually do one commit per logical change, and this feature seems small enough that it can just be one single commit.

File: internal/ircserver/cmd_join.go (hunk 1 above)
Comment (danielkim802) [reply internal/ircserver/cmd_join.go]: Missed this, fixed.

File: serialization_test.go (hunk 2 above)
Comment (danielkim802) [reply serialization_test.go:62]: I wanted to add test coverage for the snapshot.proto and serialize.go changes which I thought this accomplished but I might be mistaken.

File: internal/ircserver/cmd_mode.go [hunk 3]
```diff
@@ -114,6 +114,27 @@ func (i *IRCServer) cmdMode(s *Session, reply *Replyctx, msg *irc.Message) {
...
+				case 'k':
+					if newvalue {
+						if mode.Param == "" {
+							break
+						}
+						c.key = mode.Param
+						i.sendUser(s, reply, &irc.Message{
+							Prefix:  i.ServerPrefix,
+							Command: irc.MODE,
+							Params:  []string{channelname, "+k", mode.Param},
```
Comment (stapelberg) [inline internal/ircserver/cmd_mode.go:126]: nit: use c.key instead of mode.Param here for symmetry with the else branch

File: serialization_test.go (hunk 2 above)
Comment (stapelberg) [reply serialization_test.go:62]: Ah, I see. In general, we trust that Protobuf will serialize a new string field without problems, so we don’t need an explicit test for this. Can you revert your changes to serialization_test.go please?

**Review by stapelberg — CHANGES_REQUESTED** [review]
This looks good overall. Once you address the remaining two nits, I can merge this.

File: serialization_test.go (hunk 2 above)
Comment (danielkim802) [reply serialization_test.go]: Done.

File: internal/ircserver/cmd_mode.go (hunk 3 above)
Comment (danielkim802) [reply internal/ircserver/cmd_mode.go]: Done.

**Review by stapelberg — APPROVED** [review]
